{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "    \n",
    "    Analyses of ribosome profiling data related to the TrmD degron project\n",
    "    \n",
    "    Copyright (C) 2021  Allen Buskirk, buskirk@jhmi.edu\n",
    "    \n",
    "    This program is free software: you can redistribute it and/or modify\n",
    "    it under the terms of the GNU General Public License as published by\n",
    "    the Free Software Foundation, either version 3 of the License, or\n",
    "    (at your option) any later version.\n",
    "    This program is distributed in the hope that it will be useful,\n",
    "    but WITHOUT ANY WARRANTY; without even the implied warranty of\n",
    "    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the\n",
    "    GNU General Public License for more details.\n",
    "    You should have received a copy of the GNU General Public License\n",
    "    along with this program.  If not, see <http://www.gnu.org/licenses/>."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The autoreload extension is already loaded. To reload it, use:\n",
      "  %reload_ext autoreload\n"
     ]
    }
   ],
   "source": [
    "%load_ext autoreload\n",
    "%autoreload 2\n",
    "import os\n",
    "import ribo_utilities\n",
    "import ribo_analysis\n",
    "from BCBio import GFF\n",
    "\n",
    "FP_names = [\n",
    "    'trmD_control_FP1',\n",
    "    'trmD_degron_FP1',\n",
    "    'trmD_control_FP2',\n",
    "    'trmD_degron_FP2',\n",
    "    'trmD_WT_FP',\n",
    "    'trmD_KO_FP',\n",
    "    'trmD_control_RNAseq1',\n",
    "    'trmD_degron_RNAseq1',\n",
    "    'trmD_control_RNAseq2',\n",
    "    'trmD_degron_RNAseq2'\n",
    "    ]\n",
    "    \n",
    "TruSeq_names = [\n",
    "    'trmD_WT_RNAseq',\n",
    "    'trmD_KO_RNAseq'\n",
    "    ]\n",
    "\n",
    "inpath      = '/home/allen/data/trmD/'    \n",
    "path_script = '/home/allen/code_trmD/'\n",
    "\n",
    "paths_in = {\n",
    "    'path_fastq'     : inpath  + 'fastq/', \n",
    "    'path_gff'       : path_script + 'coli2.gff',\n",
    "    'path_bowtie'    : 'bowtie',\n",
    "    'index_ladder'   : path_script + 'indexes/ladder/ladder',\n",
    "    'index_trna'     : path_script + 'indexes/MG1655/MG1655_tRNA',\n",
    "    'index_rrna'     : path_script + 'indexes/MG1655/MG1655_rRNA',\n",
    "    'index_chr'      : path_script + 'indexes/MG1655/MG1655_genome',\n",
    "    }\n",
    "\n",
    "paths_out = {\n",
    "    'path_filter'       : inpath  + 'filterdata/',\n",
    "    'path_ladder'       : inpath  + 'alignments/ladder/',\n",
    "    'path_trna'         : inpath  + 'alignments/tRNA/',\n",
    "    'path_rrna'         : inpath  + 'alignments/rRNA/',\n",
    "    'path_chr'          : inpath  + 'alignments/chr/',\n",
    "    'path_temp'         : inpath  + 'tmpds/',\n",
    "    'path_density'      : inpath  + 'density/',\n",
    "    'path_wig'          : inpath  + 'wigfiles/',\n",
    "    'path_log'          : inpath  + 'log/',\n",
    "    'path_genelists'    : inpath  + 'analysis/genelists/',\n",
    "    'path_pauses'       : inpath  + 'analysis/pauses/'\n",
    "    }\n",
    "\n",
    "for i in paths_out:\n",
    "    if not os.path.exists(paths_out[i]):\n",
    "        os.makedirs(paths_out[i])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "log saved at paths_out['path_filter']\n",
      "trmD_control_FP1 completed\n",
      "trmD_degron_FP1 completed\n"
     ]
    }
   ],
   "source": [
    "## use skewer version 0.2.2 to remove the linker from the reads in the FASTQ files \n",
    "## for the footprints libraries and the RNAseq from the trmD-degron samples\n",
    "## which were made from 10-40 nt mRNA fragments\n",
    "\n",
    "import os\n",
    "\n",
    "minlength = 10\n",
    "maxlength = 40\n",
    "phred_cutoff = 10\n",
    "\n",
    "linker = 'CTGTAGGCACCATCAATAGATCGGAAGAGCACACGTCTGAACTCCAGTCA'\n",
    "\n",
    "files = FP_names\n",
    "\n",
    "print \"log saved at paths_out['path_filter']\"\n",
    "\n",
    "for fname in files:\n",
    "    file_in = paths_in['path_fastq'] + fname + '.fastq'\n",
    "    file_out = paths_out['path_filter'] + fname\n",
    "\n",
    "    skewer_command = 'skewer -x %s -Q %d  -l %d -L %d -o %s --quiet %s' % (\n",
    "        linker, \n",
    "        phred_cutoff, \n",
    "        minlength,\n",
    "        maxlength, \n",
    "        file_out, \n",
    "        file_in \n",
    "        )\n",
    "\n",
    "    os.system(skewer_command)\n",
    "    print fname + \" completed\"\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "log saved at paths_out['path_filter']\n",
      "trmD_WT_RNAseq completed\n",
      "trmD_KO_RNAseq completed\n"
     ]
    }
   ],
   "source": [
    "## for TruSeq, use skewer for quality filtering only, not trimming the linker\n",
    "\n",
    "import os\n",
    "\n",
    "phred_cutoff = 10\n",
    "\n",
    "files = TruSeq_names\n",
    "\n",
    "print \"log saved at paths_out['path_filter']\"\n",
    "\n",
    "for fname in files:\n",
    "    file_in = paths_in['path_fastq'] + fname + '.fastq'\n",
    "    file_out = paths_out['path_filter'] + fname\n",
    "\n",
    "    skewer_command = 'skewer -Q %d -o %s --quiet %s' % (\n",
    "        phred_cutoff, \n",
    "        file_out, \n",
    "        file_in \n",
    "        )\n",
    "\n",
    "    os.system(skewer_command)\n",
    "    print fname + \" completed\"\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "trmD_WT_RNAseq\n",
      "trmD_KO_RNAseq\n"
     ]
    }
   ],
   "source": [
    "# alignments using bowtie 1.2.2\n",
    "# this function aligns to a ladder index, then the tRNA index, then the rRNA index, then the chromosome\n",
    "# tRNA and rRNA match alignments stored in the tmpds folder\n",
    "# makes tmpds folder and alignment folders automatically\n",
    "\n",
    "import os\n",
    "import sys\n",
    "\n",
    "# files = FP_names + TruSeq_names\n",
    "files = TruSeq_names\n",
    "\n",
    "######################################################\n",
    "\n",
    "path_l = paths_out['path_ladder']\n",
    "path_t = paths_out['path_trna']\n",
    "path_r = paths_out['path_rrna']\n",
    "path_chr = paths_out['path_chr']\n",
    "path_temp = paths_out['path_temp']\n",
    "\n",
    "if os.path.exists(paths_out['path_filter']+files[0]+'_filtered_trimmed.fastq'):\n",
    "    trimmed = '_filtered_trimmed.fastq '\n",
    "elif os.path.exists(paths_out['path_filter']+files[0]+'-trimmed.fastq'):\n",
    "    trimmed = '-trimmed.fastq '\n",
    "else:\n",
    "    sys.exit(\"error: trimmed file name\")\n",
    "    \n",
    "for fname in files:\n",
    "    print fname\n",
    "    \n",
    "    #first, align to ladder index to subtract any ladder oligos that were cloned\n",
    "    bowtie_l = paths_in['path_bowtie'] + ' -v 2 -y -m 1 -a --best --strata -S -p 2 --un '\n",
    "    bowtie_l += path_l + fname + '_nomatch.fastq --max ' + path_l + fname + '_multi.fastq --al '\n",
    "    bowtie_l += path_l + fname + '_match.fastq ' + paths_in['index_ladder'] + ' '\n",
    "    bowtie_l += paths_out['path_filter'] + fname + trimmed + ' 1> ' + path_temp + fname + '_ladder_match.SAM'\n",
    "    bowtie_l += ' 2> ' + paths_out['path_log'] + fname\n",
    "    os.system(bowtie_l)\n",
    "\n",
    "    # second, align to tRNA index\n",
    "    bowtie_tRNA = paths_in['path_bowtie'] + ' -v 2 -y -m 1 -a --best --strata -S -p 2 --un '\n",
    "    bowtie_tRNA += path_t + fname + '_nomatch.fastq --max ' + path_t + fname + '_multi.fastq --al '\n",
    "    bowtie_tRNA += path_t + fname + '_match.fastq ' + paths_in['index_trna'] + ' '\n",
    "    bowtie_tRNA += path_l + fname + '_nomatch.fastq ' + ' 1> ' + path_temp + fname + '_tRNA_match.SAM'\n",
    "    bowtie_tRNA += ' 2>> ' + paths_out['path_log'] + fname \n",
    "    os.system(bowtie_tRNA)\n",
    "\n",
    "    # third, align to the rRNA index\n",
    "    bowtie_rRNA = paths_in['path_bowtie'] + ' -v 2 -y -m 1 -a --best --strata -S -p 2 --un '\n",
    "    bowtie_rRNA += path_r + fname + '_nomatch.fastq --max ' + path_r + fname + '_multi.fastq --al '\n",
    "    bowtie_rRNA += path_r + fname + '_match.fastq ' + paths_in['index_rrna'] + ' '\n",
    "    bowtie_rRNA += path_t + fname + '_nomatch.fastq ' + ' 1> ' + path_temp + fname + '_rRNA_match.SAM'\n",
    "    bowtie_rRNA += ' 2>> ' + paths_out['path_log'] + fname \n",
    "    os.system(bowtie_rRNA)\n",
    "\n",
    "    # finally align to the chr index\n",
    "    bowtie_chr = paths_in['path_bowtie'] + ' -v 2 -y -m 1 -a --best --strata -S -p 2 --un '\n",
    "    bowtie_chr += path_chr + fname + '_nomatch.fastq --max ' + path_chr + fname + '_multi.fastq --al '\n",
    "    bowtie_chr += path_chr + fname + '_match.fastq ' + paths_in['index_chr'] + ' '\n",
    "    bowtie_chr += path_r + fname + '_nomatch.fastq ' + ' 1> ' + path_chr + fname + '_match.SAM'\n",
    "    bowtie_chr += ' 2>> ' + paths_out['path_log'] + fname \n",
    "    os.system(bowtie_chr)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "-----DENSITY-----\n",
      "\n",
      "Files to condense: trmD_control_FP1, trmD_degron_FP1, trmD_control_FP2, trmD_degron_FP2, trmD_WT_FP, trmD_KO_FP, trmD_control_RNAseq1, trmD_degron_RNAseq1, trmD_control_RNAseq2, trmD_degron_RNAseq2\n",
      "\n",
      "\tStarted density at 2021-02-04 16:01:28.536678\n",
      "\ttrmD_degron_RNAseq2 3222468\n",
      "\ttrmD_control_RNAseq2 4138978\n",
      "\ttrmD_degron_RNAseq1 4061634\n",
      "\ttrmD_KO_FP 7137690\n",
      "\ttrmD_WT_FP 8648301\n",
      "\ttrmD_control_RNAseq1 8449124\n",
      "\ttrmD_degron_FP1 19918072\n",
      "\ttrmD_degron_FP2 19998980\n",
      "\ttrmD_control_FP1 20466441\n",
      "\ttrmD_control_FP2 20760261\n",
      "\tFinished density at 2021-02-04 16:02:35.301536\n",
      "\tCOMPLETED DENSITY\n"
     ]
    }
   ],
   "source": [
    "### creating ribosome density files, saving pickle file and WIG file\n",
    "\n",
    "files = FP_names\n",
    "minlength = 10\n",
    "maxlength = 40\n",
    "threads = 4\n",
    "\n",
    "ribo_utilities.run_density(files, minlength, maxlength, threads, paths_in, paths_out, 'ribo_seq')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "-----DENSITY-----\n",
      "\n",
      "Files to condense: trmD_WT_RNAseq, trmD_KO_RNAseq\n",
      "\n",
      "\tStarted density at 2021-02-05 15:25:44.922732\n",
      "\ttrmD_WT_RNAseq 19239048\n",
      "\ttrmD_KO_RNAseq 32082308\n",
      "\tFinished density at 2021-02-05 15:27:25.353326\n",
      "\tCOMPLETED DENSITY\n"
     ]
    }
   ],
   "source": [
    "### for TruSeq creating the same density files, pickle and WIG\n",
    "### adding \"TruSeq\" to the argument flips the strands\n",
    "###\n",
    "### by hand, you must collect all the output and save in the reads.csv file in the density folder with format:\n",
    "### filename,mapped_reads\n",
    "\n",
    "files = TruSeq_names\n",
    "minlength = 40\n",
    "maxlength = 60\n",
    "threads = 4\n",
    "\n",
    "ribo_utilities.run_density(files, minlength, maxlength, threads, paths_in, paths_out, 'TruSeq')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "trmD_WT_RNAseq\n",
      "Genes below threshold = 0\n",
      "Genes dropped by givegene (overlap, undesirable features, etc.) = 872\n",
      "Genes in list = 4146\n",
      "\n",
      "trmD_KO_RNAseq\n",
      "Genes below threshold = 0\n",
      "Genes dropped by givegene (overlap, undesirable features, etc.) = 872\n",
      "Genes in list = 4146\n"
     ]
    }
   ],
   "source": [
    "### generate a list of genes with the rpkm values and reads to quantitate expression\n",
    "\n",
    "files = TruSeq_names\n",
    "\n",
    "thresh = -1      # threshold value in reads per codon, -1 gives you all genes\n",
    "shift = 0        # no shift for TruSeq samples\n",
    "\n",
    "# the totalreads for each library is stored in reads.csv\n",
    "# and is needed to calculate reads per codon, since the density\n",
    "# files are stored as rpm not absolute counts\n",
    "reads_file = paths_out['path_density'] + 'reads.csv'\n",
    "f_reads = open(reads_file, 'rU')\n",
    "readlist = ribo_analysis.readindict(f_reads)\n",
    "f_reads.close()\n",
    "\n",
    "GFFgen = GFF.parse(paths_in['path_gff'])\n",
    "chrom = GFFgen.next()\n",
    "\n",
    "for fname in files:\n",
    "    print \"\"\n",
    "    print fname\n",
    "    \n",
    "    totalreads = int(readlist[fname][0])\n",
    "\n",
    "    counts0 = ribo_utilities.unPickle(paths_out['path_density'] + fname + '_plus')\n",
    "    counts1 = ribo_utilities.unPickle(paths_out['path_density'] + fname + '_minus')\n",
    "    counts = [counts0[chrom.id],counts1[chrom.id]]\n",
    "    \n",
    "    mgl = ribo_analysis.makegenelist(counts, chrom, shift, thresh, totalreads)\n",
    "    ribo_analysis.writedicttoexcel(mgl, paths_out['path_genelists'] + fname + \"_genelist\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "files = FP_names\n",
    "\n",
    "thresh = -1      # threshold value in reads per codon, -1 gives you all genes\n",
    "shift = -15      # shifting reads to line up with ribosome sites, -15 gives you the P site in E coli\n",
    "\n",
    "# the totalreads for each library is stored in reads.csv\n",
    "# and is needed to calculate reads per codon, since the density\n",
    "# files are stored as rpm not absolute counts\n",
    "reads_file = paths_out['path_density'] + 'reads.csv'\n",
    "f_reads = open(reads_file, 'rU')\n",
    "readlist = ribo_analysis.readindict(f_reads)\n",
    "f_reads.close()\n",
    "\n",
    "GFFgen = GFF.parse(paths_in['path_gff'])\n",
    "chrom = GFFgen.next()\n",
    "\n",
    "for fname in files:\n",
    "    print \"\"\n",
    "    print fname\n",
    "    \n",
    "    totalreads = int(readlist[fname][0])\n",
    "\n",
    "    counts0 = ribo_utilities.unPickle(paths_out['path_density'] + fname + '_plus')\n",
    "    counts1 = ribo_utilities.unPickle(paths_out['path_density'] + fname + '_minus')\n",
    "    counts = [counts0[chrom.id],counts1[chrom.id]]\n",
    "    \n",
    "    mgl = ribo_analysis.makegenelist(counts, chrom, shift, thresh, totalreads)\n",
    "    ribo_analysis.writedicttoexcel(mgl, paths_out['path_genelists'] + fname + \"_genelist\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "trmD_control_FP1\n",
      "trmD_degron_FP1\n",
      "trmD_control_FP2\n",
      "trmD_degron_FP2\n",
      "trmD_WT_FP\n",
      "trmD_KO_FP\n"
     ]
    }
   ],
   "source": [
    "### code for calculating ribosome pausing at the A site codon for all 20 amino acids \n",
    "### the pause score is called \"oldpause\" and is found in the csv file\n",
    "### the binary file output contains 21 plots (end-to-end) visualized with Igor Pro in the figures\n",
    "### plots in the binary file occur in the same order as the csv file\n",
    "\n",
    "files = ['trmD_control_FP1',\n",
    "    'trmD_degron_FP1',\n",
    "    'trmD_control_FP2',\n",
    "    'trmD_degron_FP2',\n",
    "        ]\n",
    "    \n",
    "motifsize = 1           # Size of the motif, in units of bp or aa as appropriate\n",
    "scorewin = [-3,-2]      # For pause scores, the region to include as density in numerator\n",
    "                        #       Scorewin[0] = 0 is the start of motif\n",
    "                        #       e.g. [0,3] gives 3 nt density: positions 0, 1, and 2\n",
    "                        #       use -3,-2 for the A site codon \n",
    "avgwin = [50,50]        # For the size of the window. [0] is nt upstream, [1] downstream\n",
    "                        #   of the start of the motif\n",
    "inframe = 2             # 0 = no frame, 1 = in frame, 2 = amino acids\n",
    "thresh = 0.1            # cutoff for inclusion for genes; gene average must be above thresh reads per codon\n",
    "shift = -15             # shifts density to line up with the ribosomal P site codon \n",
    "threads = 4\n",
    "\n",
    "# the totalreads for each library is stored in reads.csv and is needed to calculate reads per codon\n",
    "# since the density files are stored as rpm not absolute reads\n",
    "reads_file = paths_out['path_density'] + 'reads.csv'\n",
    "f_reads = open(reads_file, 'rU')\n",
    "readlist = ribo_analysis.readindict(f_reads)\n",
    "f_reads.close()\n",
    "\n",
    "GFFgen = GFF.parse(paths_in['path_gff'])\n",
    "chrom = GFFgen.next()\n",
    "\n",
    "arguments = []\n",
    "\n",
    "for fname in files:\n",
    "    print fname\n",
    "    \n",
    "    totalreads = int(readlist[fname][0])\n",
    "\n",
    "    counts0 = ribo_utilities.unPickle(paths_out['path_density'] + fname + '_plus')\n",
    "    counts1 = ribo_utilities.unPickle(paths_out['path_density'] + fname + '_minus')\n",
    "    counts = [counts0[chrom.id],counts1[chrom.id]]\n",
    "        \n",
    "    outfilestring = paths_out['path_pauses'] + fname + '_A_site_aa'\n",
    "    \n",
    "    arg = [chrom, counts, outfilestring, motifsize, inframe, thresh, totalreads, shift, avgwin, scorewin]\n",
    "    arguments.append(arg)\n",
    "\n",
    "ribo_utilities.multiprocess(ribo_analysis.motifavg_ab_wf, arguments, threads)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "-----DENSITY-----\n",
      "\n",
      "Files to condense: trmD_WT_FP, trmD_KO_FP\n",
      "\n",
      "\tStarted density at 2021-02-08 10:25:02.911870\n",
      "\ttrmD_KO_FP 4158284\n",
      "\ttrmD_WT_FP 5119259\n",
      "\tFinished density at 2021-02-08 10:25:42.247537\n",
      "\tCOMPLETED DENSITY\n"
     ]
    }
   ],
   "source": [
    "### some of the short reads in the two Trm5 samples don't really have a 3'-end \n",
    "### corresponding to the 3'-boundary of the ribosome, so remove those for pausing analyses\n",
    "### focusing only on reads 24 - 40 nt in length\n",
    "### this will create density and wig files named _long for these analyses\n",
    "\n",
    "files = ['trmD_WT_FP', 'trmD_KO_FP']\n",
    "minlength = 24\n",
    "maxlength = 40\n",
    "threads = 4\n",
    "\n",
    "ribo_utilities.run_density(files, minlength, maxlength, threads, paths_in, paths_out, 'ribo_seq')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "trmD_WT_FP_long\n",
      "trmD_KO_FP_long\n"
     ]
    }
   ],
   "source": [
    "### pausing analyses for the long reads of the trmD_KO and _WT strains\n",
    "### code for calculating ribosome pausing at the A site codon for all 20 amino acids \n",
    "### scorewin is slightly different here because of the way the reads line up \n",
    "### (these were made with a different lysis buffer)\n",
    "\n",
    "files = ['trmD_WT_FP_long', 'trmD_KO_FP_long']\n",
    "    \n",
    "motifsize = 1           # Size of the motif, in units of bp or aa as appropriate\n",
    "scorewin = [-4,-2]      # For pause scores, the region to include as density in numerator\n",
    "                        #       Scorewin[0] = 0 is the start of motif\n",
    "                        #       e.g. [0,3] gives 3 nt density: positions 0, 1, and 2\n",
    "                        #       use -3,-2 for the A site codon \n",
    "avgwin = [50,50]        # For the size of the window. [0] is nt upstream, [1] downstream\n",
    "                        #   of the start of the motif\n",
    "inframe = 2             # 0 = no frame, 1 = in frame, 2 = amino acids\n",
    "thresh = 0.1            # cutoff for inclusion for genes; gene average must be above thresh reads per codon\n",
    "shift = -15             # shifts density to line up with the ribosomal P site codon \n",
    "threads = 4\n",
    "\n",
    "# the totalreads for each library is stored in reads.csv and is needed to calculate reads per codon\n",
    "# since the density files are stored as rpm not absolute reads\n",
    "reads_file = paths_out['path_density'] + 'reads.csv'\n",
    "f_reads = open(reads_file, 'rU')\n",
    "readlist = ribo_analysis.readindict(f_reads)\n",
    "f_reads.close()\n",
    "\n",
    "GFFgen = GFF.parse(paths_in['path_gff'])\n",
    "chrom = GFFgen.next()\n",
    "\n",
    "arguments = []\n",
    "\n",
    "for fname in files:\n",
    "    print fname\n",
    "    \n",
    "    totalreads = int(readlist[fname][0])\n",
    "\n",
    "    counts0 = ribo_utilities.unPickle(paths_out['path_density'] + fname + '_plus')\n",
    "    counts1 = ribo_utilities.unPickle(paths_out['path_density'] + fname + '_minus')\n",
    "    counts = [counts0[chrom.id],counts1[chrom.id]]\n",
    "        \n",
    "    outfilestring = paths_out['path_pauses'] + fname + '_A_site_aa'\n",
    "    \n",
    "    arg = [chrom, counts, outfilestring, motifsize, inframe, thresh, totalreads, shift, avgwin, scorewin]\n",
    "    arguments.append(arg)\n",
    "\n",
    "ribo_utilities.multiprocess(ribo_analysis.motifavg_ab_wf, arguments, threads)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "trmD_control_FP1\n",
      "trmD_degron_FP1\n",
      "trmD_control_FP2\n",
      "trmD_degron_FP2\n"
     ]
    }
   ],
   "source": [
    "### code for calculating ribosome pausing at the A site codon for all 64 codons\n",
    "### the pause score is called \"oldpause\" and is found in the csv file\n",
    "### the binary file output contains 63 plots (end-to-end) visualized with Igor Pro in the figures\n",
    "\n",
    "files = ['trmD_control_FP1',\n",
    "    'trmD_degron_FP1',\n",
    "    'trmD_control_FP2',\n",
    "    'trmD_degron_FP2',\n",
    "        ]\n",
    "    \n",
    "motifsize = 3           # Size of the motif, in units of bp or aa as appropriate\n",
    "scorewin = [-3,-2]      # For pause scores, the region to include as density in numerator\n",
    "                        #       Scorewin[0] = 0 is the start of motif\n",
    "                        #       e.g. [0,3] gives 3 nt density: positions 0, 1, and 2\n",
    "                        #       use -3,-2 for the A site codon \n",
    "avgwin = [50,50]        # For the size of the window. [0] is nt upstream, [1] downstream\n",
    "                        #   of the start of the motif\n",
    "inframe = 1             # 0 = no frame, 1 = in frame, 2 = amino acids\n",
    "thresh = 0.1            # cutoff for inclusion for genes; gene average must be above thresh reads per codon\n",
    "shift = -15             # shifts density to line up with the ribosomal P site codon \n",
    "threads = 4\n",
    "\n",
    "# the totalreads for each library is stored in reads.csv and is needed to calculate reads per codon\n",
    "# since the density files are stored as rpm not absolute reads\n",
    "reads_file = paths_out['path_density'] + 'reads.csv'\n",
    "f_reads = open(reads_file, 'rU')\n",
    "readlist = ribo_analysis.readindict(f_reads)\n",
    "f_reads.close()\n",
    "\n",
    "GFFgen = GFF.parse(paths_in['path_gff'])\n",
    "chrom = GFFgen.next()\n",
    "\n",
    "arguments = []\n",
    "\n",
    "for fname in files:\n",
    "    print fname\n",
    "    \n",
    "    totalreads = int(readlist[fname][0])\n",
    "\n",
    "    counts0 = ribo_utilities.unPickle(paths_out['path_density'] + fname + '_plus')\n",
    "    counts1 = ribo_utilities.unPickle(paths_out['path_density'] + fname + '_minus')\n",
    "    counts = [counts0[chrom.id],counts1[chrom.id]]\n",
    "        \n",
    "    outfilestring = paths_out['path_pauses'] + fname + '_A_site_codons'\n",
    "    \n",
    "    arg = [chrom, counts, outfilestring, motifsize, inframe, thresh, totalreads, shift, avgwin, scorewin]\n",
    "    arguments.append(arg)\n",
    "\n",
    "ribo_utilities.multiprocess(ribo_analysis.motifavg_ab_wf, arguments, threads)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "trmD_WT_FP_long\n",
      "trmD_KO_FP_long\n"
     ]
    }
   ],
   "source": [
    "### pausing analyses for the long reads of the trmD_KO and _WT strains\n",
    "### code for calculating ribosome pausing at the A site codon for all 64 codons \n",
    "### scorewin is slightly different here because of the way the reads line up \n",
    "### (these were made with a different lysis buffer)\n",
    "\n",
    "files = ['trmD_WT_FP_long', 'trmD_KO_FP_long']\n",
    "    \n",
    "motifsize = 3           # Size of the motif, in units of bp or aa as appropriate\n",
    "scorewin = [-4,-2]      # For pause scores, the region to include as density in numerator\n",
    "                        #       Scorewin[0] = 0 is the start of motif\n",
    "                        #       e.g. [0,3] gives 3 nt density: positions 0, 1, and 2\n",
    "                        #       use -3,-2 for the A site codon \n",
    "avgwin = [50,50]        # For the size of the window. [0] is nt upstream, [1] downstream\n",
    "                        #   of the start of the motif\n",
    "inframe = 1             # 0 = no frame, 1 = in frame, 2 = amino acids\n",
    "thresh = 0.1            # cutoff for inclusion for genes; gene average must be above thresh reads per codon\n",
    "shift = -15             # shifts density to line up with the ribosomal P site codon \n",
    "threads = 4\n",
    "\n",
    "# the totalreads for each library is stored in reads.csv and is needed to calculate reads per codon\n",
    "# since the density files are stored as rpm not absolute reads\n",
    "reads_file = paths_out['path_density'] + 'reads.csv'\n",
    "f_reads = open(reads_file, 'rU')\n",
    "readlist = ribo_analysis.readindict(f_reads)\n",
    "f_reads.close()\n",
    "\n",
    "GFFgen = GFF.parse(paths_in['path_gff'])\n",
    "chrom = GFFgen.next()\n",
    "\n",
    "arguments = []\n",
    "\n",
    "for fname in files:\n",
    "    print fname\n",
    "    \n",
    "    totalreads = int(readlist[fname][0])\n",
    "\n",
    "    counts0 = ribo_utilities.unPickle(paths_out['path_density'] + fname + '_plus')\n",
    "    counts1 = ribo_utilities.unPickle(paths_out['path_density'] + fname + '_minus')\n",
    "    counts = [counts0[chrom.id],counts1[chrom.id]]\n",
    "        \n",
    "    outfilestring = paths_out['path_pauses'] + fname + '_A_site_codons'\n",
    "    \n",
    "    arg = [chrom, counts, outfilestring, motifsize, inframe, thresh, totalreads, shift, avgwin, scorewin]\n",
    "    arguments.append(arg)\n",
    "\n",
    "ribo_utilities.multiprocess(ribo_analysis.motifavg_ab_wf, arguments, threads)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
